lambda/lambda.ml

```diff
@@ -469,6 +469,32 @@ let name_lambda_list args fn =
       Llet(Strict, Pgenval, id, arg, name_list (Lvar id :: names) rem) in
   name_list [] args
 
+let collapse_apply
+    ({ ap_loc; ap_func; ap_args;
+       ap_tailcall=_; ap_inlined=_; ap_specialised=_; }
+     as apply) =
+  (* Note: We're dropping the attributes specific to this application
+     if we merge it with another application.
+     But consider this case:
+
+     let f x = f_body in
+     let g x = do_something (); f in
+     (((g @[inlined]) a) [@inlined]) b
+
+     Here it seems legitimate to put both annotations, and it seems wrong to
+     discard the second one.
+     So it might be better to never collapse applications if the outer one has
+     non-default attributes. *)
+  match ap_func with
+  | Lsend(k, lmet, lobj, largs, _) ->
+      Lsend(k, lmet, lobj, largs @ ap_args, ap_loc)
+  | Levent(Lsend(k, lmet, lobj, largs, _), _) ->
+      Lsend(k, lmet, lobj, largs @ ap_args, ap_loc)
+  | Lapply ap ->
+      Lapply {ap with ap_args = ap.ap_args @ ap_args; ap_loc}
```

Comment:
Should we emit a misplaced attribute warning rather than silently dropping the outer attribute? 
The fact that
```ocaml
let rec f x y = 1 + (f x[@tailcall]) y
```
compiles without any warning seems misleading.

Reply:
Indeed (although you do get a warning if you compile in bytecode with `-g`).
As far as I can tell this was already the case in 4.12, and although I haven't checked I wouldn't be surprised if it's been that way since the introduction of the `@tailcall` attribute.
Do you want me to try to improve the situation in this PR, or do you prefer to keep this one as a simple revert to the 4.12 situation and propose additional fixes in other PRs ?
(Either way, I will be away for at least a week, so don't hesitate to push changes yourself if you need to make progress quickly)

Reply:
We can indeed focus on reverting to the 4.12 situation for (n>1) application mixed with `@@` and `|>` first.